Sources/AsyncHTTPClient/ConnectionPool/HTTPConnectionPool+Factory.swift

```diff
@@ -361,14 +361,19 @@ extension HTTPConnectionPool.ConnectionFactory {
         var channelFuture = bootstrapFuture.flatMap { bootstrap -> EventLoopFuture<Channel> in
             return bootstrap.connect(target: self.key.connectionTarget)
         }.flatMap { channel -> EventLoopFuture<(Channel, String?)> in
-            // It is save to use `try!` here, since we are sure, that a `TLSEventsHandler` exists
-            // within the pipeline. It is added in `makeTLSBootstrap`.
-            let tlsEventHandler = try! channel.pipeline.syncOperations.handler(type: TLSEventsHandler.self)
-
-            // The tlsEstablishedFuture is set as soon as the TLSEventsHandler is in a
-            // pipeline. It is created in TLSEventsHandler's handlerAdded method.
-            return tlsEventHandler.tlsEstablishedFuture!.flatMap { negotiated in
-                channel.pipeline.removeHandler(tlsEventHandler).map { (channel, negotiated) }
+            do {
+                // if the channel is closed before flatMap is executed, all ChannelHandler are removed
+                // and TLSEventsHandler is therefore not present either
+                let tlsEventHandler = try channel.pipeline.syncOperations.handler(type: TLSEventsHandler.self)
+                
+                // The tlsEstablishedFuture is set as soon as the TLSEventsHandler is in a
+                // pipeline. It is created in TLSEventsHandler's handlerAdded method.
+                return tlsEventHandler.tlsEstablishedFuture!.flatMap { negotiated in
+                    channel.pipeline.removeHandler(tlsEventHandler).map { (channel, negotiated) }
+                }
+            } catch {
+                precondition(channel.isActive == false, "if the channel is still active then TLSEventsHandler must be present but got error \(error)")
```

Comment:
I don't think this precondition is particularly wise: the connection pool will appropriately handle this. Let's drop it to `assert` instead.

Reply:
Done in [d0d1cee](https://github.com/swift-server/async-http-client/pull/671/commits/d0d1cee09e56b1c34758b270a5da99114903b6cc)